source/multigrid/mg_transfer_matrix_free.cc

```diff
@@ -296,14 +296,31 @@ namespace
           evaluator.template values<0,prolongate,false>(t0, t2);
         else if (dim == 2)
           {
-            evaluator.template values<0,prolongate,false>(t0, t1);
-            evaluator.template values<1,prolongate,false>(t1, t2);
+            if (prolongate)
+              {
+                evaluator.template values<0,true,false>(t0, t1);
+                evaluator.template values<1,true,false>(t1, t2);
+              }
+            else
+              {
+                evaluator.template values<1,false,false>(t0, t1);
+                evaluator.template values<0,false,false>(t1, t2);
+              }
           }
         else if (dim == 3)
           {
-            evaluator.template values<0,prolongate,false>(t0, t2);
-            evaluator.template values<1,prolongate,false>(t2, t1);
-            evaluator.template values<2,prolongate,false>(t1, t2);
+            if (prolongate)
+              {
+                evaluator.template values<0,true,false>(t0, t2);
+                evaluator.template values<1,true,false>(t2, t1);
+                evaluator.template values<2,true,false>(t1, t2);
+              }
+            else
+              {
+                evaluator.template values<2,false,false>(t0, t2);
```

Comment:
can you please leave a comment to explain this reverse ordering? (also above)

Reply:
It is because the `EvaluatorTensorProduct` kernels have gained the possibility to do in-place operations, which necessitated a change in the hardcoded strides/offsets in the generalized matrix-matrix multiplication to make sure it does not overwrite things. In particular, we need to start with the outermost dimension in the no-transpose path with more columns than rows (second `false` template argument). By the way, this reminds me that we could now do all this in a single array `t0` without temporary arrays `t1,t2`.

Reply:
We have replaced the whole function by the new `FEEvaluationImplBasisChange::do_forward/do_backward` call.